web_monitoring/diffing_server.py

```diff
@@ -226,7 +226,14 @@ def fetch_diffable_content(self, url, expected_hash, query_params):
 
             try:
                 response = yield client.fetch(url, headers=headers,
-                                              validate_cert=VALIDATE_TARGET_CERTIFICATES)
+                                              validate_cert=VALIDATE_TARGET_CERTIFICATES,
+                                              raise_error=False)
+                # Accept response with HTTP status >= 300 only if its coming
+                # from a web archive. HTTP header `Memento-Datetime` is an
+                # indication about that.
+                if response.code >= 300 and \
+                        response.headers.get('Memento-Datetime') is None:
+                    response.rethrow()
```

Comment:
`fetch_diffable_content` returns `None` to signal that it failed. However, it depends on the fact that, if `client.fetch()` raised an error `response` will be `None`:

https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/ea6f992b2a6639461ab45667b177d860724f9ebd/web_monitoring/diffing_server.py#L248

And in the calling function, we do `if not all(content)` to check if either fetch returned `None`:

https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/ea6f992b2a6639461ab45667b177d860724f9ebd/web_monitoring/diffing_server.py#L179-L184

However, now you’re populating `response`, so the calling function doesn’t realize that the request failed and so it keeps trying to do a diff anyway.

So… there are probably a few options here:

- Store `response` in a different variable so you can call `rethrow()` on it:

    ```py
    bad_response = response
    response = None
    bad_response.rethrow()
    ```

- Don’t use `raise_error=False` and instead pull the response object out of the error:

    ```py
    try:
        response = yield client.fetch(url, headers=headers,
                                      validate_cert=VALIDATE_TARGET_CERTIFICATES)
    except SomeOtherErrorType:
        ...
    except tornado.httpclient.HTTPError as error:
        # If the response is actually coming from a web archive,
        # allow error codes. The Memento-Datetime header indicates
        # the response is an archived one, and not an actual failure
        # to respond with the desired content.
        if error.response is not None and error.response.headers.get('Memento-Datetime') is not None:
            response = error.response
        else:
            self.send_error(502,
                            reason=f'Received a {error.response.code} status while fetching "{url}": {error}')
    ```

    Upside: no need to check status codes. Downside: this code is definitely a little less clear.

- Change the error handling in some more significant way. (After re-reading the code here, I realize we need to do something like that anyhow, since both the fetches could have errors, and we have similar issues. BUT we can also leave fixing that for a separate PR.)

Reply:
Thank you for taking the time to explain this so well.

I prefer your 2nd suggestion. It is much more elegant IMHO. I've updated my PR following that approach.